source/extensions/filters/network/thrift_proxy/metadata.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <string.h>
+
+#include <algorithm>
+#include <list>
+#include <string>
+
+#include "common/common/macros.h"
+
+#include "extensions/filters/network/thrift_proxy/thrift.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+/**
+ * Header is a name-value pair in Thrift transport or protocol headers.
+ */
+class Header {
+public:
+  Header(const std::string key, const std::string value) : key_(key), value_(value) {}
+  Header(const Header& rhs) : key_(rhs.key_), value_(rhs.value_) {}
+
+  const std::string& key() const { return key_; }
+  const std::string& value() const { return value_; }
+
+private:
+  std::string key_;
+  std::string value_;
+};
+
+/*
+ * HeaderMap contains Thrift transport and/or protocol-level headers.
+ */
+class HeaderMap {
```

Comment:
I considered using the Http::HeaderMap. I like the idea of reusing LowerCaseString and HeaderEntryImpl and it's memory management. I balked because it seems likely that most of the 70-odd inline headers will never see use in Thrift. (The non-inline headers are stored in a list, as in the local implementation). I suppose using the Http::HeaderMap will make my life easier when I add logging, since that API takes them. I'll have another look at reusing it.

Reply:
FWIW, at some point I would like to make the inline headers that the HeaderMapImpl supports runtime configurable. I think this is possible with some work. Obviously not suggesting you do that, but I thought I would throw that out there. I think that's potentially interesting when private filters want to add extra inline headers and/or we would like to prune some of the inline headers that might not be used in all installations.

Reply:
That would be awesome.

@htuch do you mind if I leave this as-is for now and reconsider it when we're actually able to use headers for routing?